proposals/distributed-async-epoch-reward.md

```diff
@@ -0,0 +1,223 @@
+account rewards have been distributed, a `RewardHistoryEntry` will be added to
+the `EpochRewardHistory` array with the total amount of reward, a distributed
+value equal to the sum of the vote account rewards. If the length of the history
+array exceeds the maximum depth (512), the oldest entries will be removed from
+the array (i.e., "popped").
+
+
+### `EpochRewardReserve` Sysvar Account
+
+The address for a particular `EpochRewardReserve` account is derived from the
+base Sysvar RewardReserve address, by hashing
+`SysvarRewardReserve11111111111111111111111111111111` with the u64 block height
+for the corresponding reward distribution block during the reward
+distribution phase.
+
+```
+crate::declare_sysvar_id!("SysvarRewardReserve11111111111111111111111111111", EpochRewardReserve);
+
+struct EpochRewardReserve {
+   rewards_hash: Hash,
+}
+```
+
+The account balance of `RewardsReserve` is equal to the total amount (in lamports) of
+all rewards to be distributed in a block. The `rewards_hash` is computed from the rewards in the block as follows.
+
+```
+rewards_hash = hash([(stake_account_key, reward_amount)])
+```
+
+At the end of reward computation phase, M `EpochRewardsReserve` sysvar
+accounts are updated, where `M` corresponds to how many blocks will be in
+the reward distribution phase, which is described in the next section.
+
+
+### Reward Distribution
+
+Reward distribution phase happens right after reward computation phase and
+lasts for `M` blocks.
+
+Based on current validator client's performance, 64 additional accounts can be
+stored per tick time without much performance impact for the overall block
+time. Given that each block time is 64 ticks time long, a total of 4,096
+accounts can be distributed per block. The total number of blocks needed tto
+distributed rewards is
+
+```
+M = ((4096 - 1)+num_stake_accounts)/4096
+```
+
+During the reward distribution phase, for the block at `block_height`, the rewards will be distributed from the
+`EpochRewardReserve` sysvar account with the address of
+"hash(`SysvarRewardReserve11111111111111111111111111111111`, block_height). As
+the stake_accounts receive their rewards from the reserve account, the balance
+of the EpochRewardReserve will decrease. Once all rewards have been
+distributed, the balance of the EpochRewardReserve account should be reduced to 0.
+
+Before each reward distribution, the `EpochRewardReserve` account's balance
+is checked to make sure there is enough balance to distribute the reward. After a
+reward is distributed, the balance in `EpochRewardReserve` account is decreased
+by the amount of the reward that was distributed. The hash of reward_account
+pubkey and reward amount is accumulated when rewards are distributed. After all
+the rewards are distributed from the reserve account for the block, the hash
+computed from all the rewards distribution is verified against the hash stored
+in the `EpochRewardReserve` account to make sure they match.
```

Comment:
Does it make sense to (re?)introduce a "root hash" in `RewardHistoryEntry` that's the hash of all the individual `EpochRewardReserve` hashes? 

There'd then be a "root accumulator hash" that starts empty and each `RewardHistoryEntry` hash is added to it. Once the "root accumulator hash" equals the "root hash", reward distribution should be complete and we can assert no future distributions through the remainder of the current epoch.

This would just be another safely check to ensure distributions are being made as intended.